File: discord_bot.js
const fs = require('fs');
const path = require('path');
const express = require('express');
const { Client, GatewayIntentBits, Partials, ChannelType, SlashCommandBuilder } = require('discord.js');

const disablePriv = process.env.DISABLE_PRIVILEGED_INTENTS === '1';
const intents = [GatewayIntentBits.Guilds, GatewayIntentBits.GuildMessages, GatewayIntentBits.MessageContent];
if (!disablePriv) {
  intents.push(GatewayIntentBits.GuildMembers, GatewayIntentBits.GuildPresences);
}

const client = new Client({ intents, partials: [Partials.Channel, Partials.Message, Partials.GuildMember, Partials.User] });
let settingsStore = {};
const commandsUsed = new Map();
const bannedWords = ['spamlink.com'];
const settingsPath = path.join(__dirname, 'settings.json');

function loadSettings() {
  if (fs.existsSync(settingsPath)) {
    try {
      settingsStore = JSON.parse(fs.readFileSync(settingsPath, 'utf-8'));
    } catch (_) {
      settingsStore = {};
    }
  }
}

function saveSettings() {
  try {
    fs.writeFileSync(settingsPath, JSON.stringify(settingsStore, null, 2), 'utf-8');
  } catch (_) {}
}

function defaultSettings() {
  return {
    welcome: { enabled: false, channel_id: null, message: '{user}님 환영합니다!' },
    moderation: { enabled: true, log_channel_id: null, auto_mod: false },
    music: { volume: 50, dj_role_id: null }
  };
}

async function sendModLog(guild, content) {
  const cfg = settingsStore[String(guild.id)] || {};
  const mod = cfg.moderation || {};
  const chId = mod.log_channel_id;
  if (!chId) return;
  const ch = guild.channels.cache.get(String(chId)) || null;
  if (!ch) return;
  try { await ch.send(content); } catch (_) {}
}

client.on('ready', async () => {
  const cmds = [
    new SlashCommandBuilder().setName('핑').setDescription('핑'),
    new SlashCommandBuilder().setName('echo').setDescription('메시지 에코').addStringOption(o=>o.setName('message').setDescription('메시지').setRequired(true)),
    new SlashCommandBuilder().setName('대시보드').setDescription('대시보드 링크'),
    new SlashCommandBuilder().setName('serverinfo').setDescription('서버 정보'),
    new SlashCommandBuilder().setName('userinfo').setDescription('유저 정보').addUserOption(o=>o.setName('member').setDescription('유저')),
    new SlashCommandBuilder().setName('clear').setDescription('메시지 삭제').addIntegerOption(o=>o.setName('amount').setDescription('수량(1-100)').setRequired(true)),
    new SlashCommandBuilder().setName('kick').setDescription('멤버 추방').addUserOption(o=>o.setName('member').setDescription('유저').setRequired(true)).addStringOption(o=>o.setName('reason').setDescription('사유')),
    new SlashCommandBuilder().setName('ban').setDescription('멤버 차단').addUserOption(o=>o.setName('member').setDescription('유저').setRequired(true)).addStringOption(o=>o.setName('reason').setDescription('사유')),
    new SlashCommandBuilder().setName('poll').setDescription('투표 생성').addStringOption(o=>o.setName('question').setDescription('질문').setRequired(true)),
    new SlashCommandBuilder().setName('설정').setDescription('현재 설정 확인'),
    new SlashCommandBuilder().setName('warn').setDescription('유저 경고').addUserOption(o=>o.setName('member').setDescription('유저').setRequired(true)).addStringOption(o=>o.setName('reason').setDescription('사유')),
    new SlashCommandBuilder().setName('mute').setDescription('유저 뮤트').addUserOption(o=>o.setName('member').setDescription('유저').setRequired(true)).addIntegerOption(o=>o.setName('minutes').setDescription('분').setRequired(false)),
    new SlashCommandBuilder().setName('unmute').setDescription('유저 언뮤트').addUserOption(o=>o.setName('member').setDescription('유저').setRequired(true)),
    new SlashCommandBuilder().setName('slowmode').setDescription('슬로우모드 설정').addIntegerOption(o=>o.setName('seconds').setDescription('초').setRequired(true)),
    new SlashCommandBuilder().setName('purgeuser').setDescription('유저 메시지 삭제').addUserOption(o=>o.setName('member').setDescription('유저').setRequired(true)).addIntegerOption(o=>o.setName('amount').setDescription('수량').setRequired(false)),
    new SlashCommandBuilder().setName('announce').setDescription('공지 전송').addStringOption(o=>o.setName('message').setDescription('메시지').setRequired(true))
  ].map(c=>c.toJSON());
  try { await client.application.commands.set(cmds); } catch (_) {}
  console.log(`${client.user.tag} connected`);
});

client.on('guildMemberAdd', async member => {
  const cfg = settingsStore[String(member.guild.id)];
  if (!cfg) return;
  const w = cfg.welcome;
  if (!w || !w.enabled) return;
  const chId = w.channel_id;
  const msg = w.message || '';
  if (!chId) return;
  const ch = member.guild.channels.cache.get(String(chId));
  if (!ch) return;
  try { await ch.send(msg.replace('{user}', `<@${member.id}>`)); } catch (_) {}
});

client.on('messageCreate', async msg => {
  if (!msg.guild || msg.author.bot) return;
  const gidAuto = String(msg.guild.id);
  const cfgAuto = settingsStore[gidAuto] || defaultSettings();
  if (cfgAuto.moderation && cfgAuto.moderation.auto_mod) {
    const lowered = msg.content.toLowerCase();
    if (bannedWords.some(w=>lowered.includes(w.toLowerCase()))) {
      try { await msg.delete(); } catch (_) {}
      await sendModLog(msg.guild, `AUTOMOD delete by ${msg.author} in ${msg.channel}`);
      return;
    }
  }
  const prefix = '!';
  if (!msg.content.startsWith(prefix)) return;
  const [cmd, ...rest] = msg.content.slice(prefix.length).trim().split(/\s+/);
  const gid = String(msg.guild.id);
  commandsUsed.set(gid, (commandsUsed.get(gid) || 0) + 1);
  if (cmd === 'ping') { await msg.reply('Pong!'); }
  else if (cmd === 'echo') { await msg.reply(rest.join(' ')); }
  else if (cmd === '대시보드' || cmd === 'dashboard') { await msg.reply('https://plugmarket.r-e.kr/'); }
  else if (cmd === 'serverinfo') {
    const g = msg.guild;
    await msg.reply(`서버: ${g.name} 채널: ${g.channels.cache.size} 역할: ${g.roles.cache.size}`);
  } else if (cmd === 'userinfo') {
    const m = msg.mentions.members.first() || msg.member;
    await msg.reply(`유저: ${m.displayName} ID: ${m.id}`);
  } else if (cmd === 'clear') {
    const n = Math.max(1, Math.min(100, parseInt(rest[0] || '0')));
    if (!msg.member.permissions.has('ManageMessages')) { await msg.reply('권한이 부족합니다'); return; }
    const deleted = await msg.channel.bulkDelete(n, true).catch(()=>null);
    await msg.reply(`삭제: ${deleted ? deleted.size : 0}`);
    await sendModLog(msg.guild, `CLEAR ${deleted ? deleted.size : 0} by ${msg.author} in ${msg.channel}`);
  } else if (cmd === 'kick') {
    if (!msg.member.permissions.has('KickMembers')) { await msg.reply('권한이 부족합니다'); return; }
    const target = msg.mentions.members.first();
    const reason = rest.slice(1).join(' ') || null;
    if (!target) { await msg.reply('대상 유저가 필요합니다'); return; }
    try { await target.kick(reason); await msg.reply(`${target} 추방`); await sendModLog(msg.guild, `KICK ${target} by ${msg.author} reason: ${reason || '-'}`); } catch (_) { await msg.reply('오류가 발생했습니다'); }
  } else if (cmd === 'ban') {
    if (!msg.member.permissions.has('BanMembers')) { await msg.reply('권한이 부족합니다'); return; }
    const target = msg.mentions.members.first();
    const reason = rest.slice(1).join(' ') || null;
    if (!target) { await msg.reply('대상 유저가 필요합니다'); return; }
    try { await target.ban({ reason }); await msg.reply(`${target} 차단`); await sendModLog(msg.guild, `BAN ${target} by ${msg.author} reason: ${reason || '-'}`); } catch (_) { await msg.reply('오류가 발생했습니다'); }
  } else if (cmd === 'poll') {
    const q = rest.join(' ');
    const m = await msg.channel.send(`투표: ${q}`);
    try { await m.react('👍'); await m.react('👎'); } catch (_) {}
  } else if (cmd === '설정') {
    const s = settingsStore[String(msg.guild.id)] || defaultSettings();
    const lines = [];
    lines.push(`환영: ${s.welcome.enabled ? 'ON' : 'OFF'} 채널: ${s.welcome.channel_id || '-'}`);
    lines.push(`모더레이션: ${s.moderation.enabled ? 'ON' : 'OFF'} 자동: ${s.moderation.auto_mod ? 'ON' : 'OFF'} 로그: ${s.moderation.log_channel_id || '-'}`);
    lines.push(`음악: 볼륨 ${s.music.volume}% DJ 역할: ${s.music.dj_role_id || '-'}`);
    await msg.reply(lines.join('\n'));
  } else if (cmd === 'warn') {
    if (!msg.member.permissions.has('KickMembers')) { await msg.reply('권한이 부족합니다'); return; }
    const target = msg.mentions.members.first();
    const reason = rest.slice(1).join(' ') || '-';
    if (!target) { await msg.reply('대상 유저가 필요합니다'); return; }
    await msg.reply(`${target} 경고: ${reason}`);
    await sendModLog(msg.guild, `WARN ${target} by ${msg.author} reason: ${reason}`);
  } else if (cmd === 'mute') {
    if (!msg.member.permissions.has('MuteMembers')) { await msg.reply('권한이 부족합니다'); return; }
    const target = msg.mentions.members.first();
    const minutes = parseInt(rest[1] || '0');
    if (!target) { await msg.reply('대상 유저가 필요합니다'); return; }
    const roleId = await ensureMuteRole(msg.guild);
    if (!roleId) { await msg.reply('Mute 역할을 만들 수 없습니다'); return; }
    try { await target.roles.add(roleId); } catch (_) { await msg.reply('역할 부여 실패'); return; }
    await msg.reply(`${target} 뮤트`);
    await sendModLog(msg.guild, `MUTE ${target} by ${msg.author} ${minutes>0?minutes+'m':''}`);
    if (minutes>0) setTimeout(async ()=>{ try { await target.roles.remove(roleId); await sendModLog(msg.guild, `UNMUTE ${target}`); } catch (_) {} }, minutes*60*1000);
  } else if (cmd === 'unmute') {
    if (!msg.member.permissions.has('MuteMembers')) { await msg.reply('권한이 부족합니다'); return; }
    const target = msg.mentions.members.first();
    if (!target) { await msg.reply('대상 유저가 필요합니다'); return; }
    const roleId = await ensureMuteRole(msg.guild);
    try { await target.roles.remove(roleId); await msg.reply(`${target} 언뮤트`); await sendModLog(msg.guild, `UNMUTE ${target} by ${msg.author}`); } catch (_) { await msg.reply('역할 제거 실패'); }
  } else if (cmd === 'slowmode') {
    if (!msg.member.permissions.has('ManageChannels')) { await msg.reply('권한이 부족합니다'); return; }
    const secs = Math.max(0, Math.min(21600, parseInt(rest[0]||'0')));
    try { await msg.channel.setRateLimitPerUser(secs); await msg.reply(`슬로우모드 ${secs}s`); } catch (_) { await msg.reply('설정 실패'); }
  } else if (cmd === 'purgeuser') {
    if (!msg.member.permissions.has('ManageMessages')) { await msg.reply('권한이 부족합니다'); return; }
    const target = msg.mentions.members.first();
    const n = Math.max(1, Math.min(100, parseInt(rest[1]||'10')));
    if (!target) { await msg.reply('대상 유저가 필요합니다'); return; }
    const messages = await msg.channel.messages.fetch({ limit: 100 }).catch(()=>null);
    const toDelete = messages ? messages.filter(m=>m.author.id===target.id).first(n) : [];
    let count = 0;
    for (const m of toDelete) { try { await m.delete(); count++; } catch (_) {} }
    await msg.reply(`유저 메시지 삭제: ${count}`);
    await sendModLog(msg.guild, `PURGEUSER ${count} of ${target} by ${msg.author} in ${msg.channel}`);
  } else if (cmd === 'announce') {
    const content = rest.join(' ');
    if (!content) { await msg.reply('메시지가 필요합니다'); return; }
    await msg.channel.send(`공지: ${content}`);
  }
});

client.on('interactionCreate', async interaction => {
  if (!interaction.isChatInputCommand()) return;
  const n = interaction.commandName;
  const gid = String(interaction.guild.id);
  commandsUsed.set(gid, (commandsUsed.get(gid) || 0) + 1);
  if (n === '핑') { await interaction.reply('Pong!'); }
  else if (n === 'echo') { await interaction.reply(interaction.options.getString('message')); }
  else if (n === '대시보드') { await interaction.reply('https://plugmarket.r-e.kr/'); }
  else if (n === 'serverinfo') {
    const g = interaction.guild;
    await interaction.reply(`서버: ${g.name} 채널: ${g.channels.cache.size} 역할: ${g.roles.cache.size}`);
  } else if (n === 'userinfo') {
    const m = interaction.options.getMember('member') || interaction.member;
    await interaction.reply(`유저: ${m.displayName} ID: ${m.id}`);
  } else if (n === 'clear') {
    const amt = Math.max(1, Math.min(100, interaction.options.getInteger('amount')));
    if (!interaction.memberPermissions.has('ManageMessages')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const deleted = await interaction.channel.bulkDelete(amt, true).catch(()=>null);
    await interaction.reply(`삭제: ${deleted ? deleted.size : 0}`);
    await sendModLog(interaction.guild, `CLEAR ${deleted ? deleted.size : 0} by ${interaction.user} in ${interaction.channel}`);
  } else if (n === 'kick') {
    if (!interaction.memberPermissions.has('KickMembers')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const target = interaction.options.getMember('member');
    const reason = interaction.options.getString('reason') || null;
    try { await target.kick(reason); await interaction.reply(`${target} 추방`); await sendModLog(interaction.guild, `KICK ${target} by ${interaction.user} reason: ${reason || '-'}`); } catch (_) { await interaction.reply({ content: '오류가 발생했습니다', ephemeral: true }); }
  } else if (n === 'ban') {
    if (!interaction.memberPermissions.has('BanMembers')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const target = interaction.options.getMember('member');
    const reason = interaction.options.getString('reason') || null;
    try { await target.ban({ reason }); await interaction.reply(`${target} 차단`); await sendModLog(interaction.guild, `BAN ${target} by ${interaction.user} reason: ${reason || '-'}`); } catch (_) { await interaction.reply({ content: '오류가 발생했습니다', ephemeral: true }); }
  } else if (n === 'poll') {
    const q = interaction.options.getString('question');
    const m = await interaction.channel.send(`투표: ${q}`);
    try { await m.react('👍'); await m.react('👎'); } catch (_) {}
    await interaction.reply({ content: '투표 생성', ephemeral: true });
  } else if (n === '설정') {
    const s = settingsStore[String(interaction.guild.id)] || defaultSettings();
    const lines = [];
    lines.push(`환영: ${s.welcome.enabled ? 'ON' : 'OFF'} 채널: ${s.welcome.channel_id || '-'}`);
    lines.push(`모더레이션: ${s.moderation.enabled ? 'ON' : 'OFF'} 자동: ${s.moderation.auto_mod ? 'ON' : 'OFF'} 로그: ${s.moderation.log_channel_id || '-'}`);
    lines.push(`음악: 볼륨 ${s.music.volume}% DJ 역할: ${s.music.dj_role_id || '-'}`);
    await interaction.reply({ content: lines.join('\n'), ephemeral: true });
  } else if (n === 'warn') {
    if (!interaction.memberPermissions.has('KickMembers')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const target = interaction.options.getMember('member');
    const reason = interaction.options.getString('reason') || '-';
    await interaction.reply(`${target} 경고: ${reason}`);
    await sendModLog(interaction.guild, `WARN ${target} by ${interaction.user} reason: ${reason}`);
  } else if (n === 'mute') {
    if (!interaction.memberPermissions.has('MuteMembers')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const target = interaction.options.getMember('member');
    const minutes = interaction.options.getInteger('minutes') || 0;
    const roleId = await ensureMuteRole(interaction.guild);
    if (!roleId) { await interaction.reply({ content: 'Mute 역할을 만들 수 없습니다', ephemeral: true }); return; }
    try { await target.roles.add(roleId); } catch (_) { await interaction.reply({ content: '역할 부여 실패', ephemeral: true }); return; }
    await interaction.reply(`${target} 뮤트`);
    await sendModLog(interaction.guild, `MUTE ${target} by ${interaction.user} ${minutes>0?minutes+'m':''}`);
    if (minutes>0) setTimeout(async ()=>{ try { await target.roles.remove(roleId); await sendModLog(interaction.guild, `UNMUTE ${target}`); } catch (_) {} }, minutes*60*1000);
  } else if (n === 'unmute') {
    if (!interaction.memberPermissions.has('MuteMembers')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const target = interaction.options.getMember('member');
    const roleId = await ensureMuteRole(interaction.guild);
    try { await target.roles.remove(roleId); await interaction.reply(`${target} 언뮤트`); await sendModLog(interaction.guild, `UNMUTE ${target} by ${interaction.user}`); } catch (_) { await interaction.reply({ content: '역할 제거 실패', ephemeral: true }); }
  } else if (n === 'slowmode') {
    if (!interaction.memberPermissions.has('ManageChannels')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const secs = Math.max(0, Math.min(21600, interaction.options.getInteger('seconds')));
    try { await interaction.channel.setRateLimitPerUser(secs); await interaction.reply(`슬로우모드 ${secs}s`); } catch (_) { await interaction.reply({ content: '설정 실패', ephemeral: true }); }
  } else if (n === 'purgeuser') {
    if (!interaction.memberPermissions.has('ManageMessages')) { await interaction.reply({ content: '권한이 부족합니다', ephemeral: true }); return; }
    const target = interaction.options.getMember('member');
    const n = Math.max(1, Math.min(100, interaction.options.getInteger('amount')||10));
    const messages = await interaction.channel.messages.fetch({ limit: 100 }).catch(()=>null);
    const toDelete = messages ? messages.filter(m=>m.author.id===target.id).first(n) : [];
    let count = 0;
    for (const m of toDelete) { try { await m.delete(); count++; } catch (_) {} }
    await interaction.reply(`유저 메시지 삭제: ${count}`);
    await sendModLog(interaction.guild, `PURGEUSER ${count} of ${target} by ${interaction.user} in ${interaction.channel}`);
  } else if (n === 'announce') {
    const content = interaction.options.getString('message');
    if (!content) { await interaction.reply({ content: '메시지가 필요합니다', ephemeral: true }); return; }
    await interaction.channel.send(`공지: ${content}`);
    await interaction.reply({ content: '공지 전송', ephemeral: true });
  }
});

loadSettings();

const app = express();
app.use(express.json());
app.use(express.static(__dirname));

app.get('/api/auth/session', (req, res) => {
  const user = { id: '0', username: 'Dashboard Admin', global_name: 'Dashboard Admin', avatar: null };
  const guilds = client.guilds.cache.map(g => ({ id: String(g.id), name: g.name, icon: g.icon, approximate_member_count: null }));
  res.json({ user, guilds, access_token: 'dev' });
});

app.post('/api/auth/discord/callback', (req, res) => {
  const user = { id: '0', username: 'Dashboard Admin', global_name: 'Dashboard Admin', avatar: null };
  const guilds = client.guilds.cache.map(g => ({ id: String(g.id), name: g.name, icon: g.icon, approximate_member_count: null }));
  res.json({ user, guilds, access_token: 'dev' });
});

app.post('/api/auth/logout', (req, res) => res.json({ ok: true }));

function getGuildOr404(id, res) {
  const g = client.guilds.cache.get(String(id));
  if (!g) { res.status(404).end(); return null; }
  return g;
}

app.get('/api/guilds/:guild_id/stats', async (req, res) => {
  const g = getGuildOr404(req.params.guild_id, res); if (!g) return;
  try { await g.members.fetch(); } catch (_) {}
  const members = g.memberCount;
  let online = 0;
  g.members.cache.forEach(m => {
    const st = m.presence ? m.presence.status : null;
    if (st === 'online' || st === 'idle' || st === 'dnd') online += 1;
  });
  const channels = g.channels.cache.size;
  const commands = commandsUsed.get(String(g.id)) || 0;
  res.json({ members, online, channels, commands });
});

app.get('/api/guilds/:guild_id/settings', (req, res) => {
  const id = String(req.params.guild_id);
  let s = settingsStore[id];
  if (!s) { s = defaultSettings(); settingsStore[id] = s; saveSettings(); }
  res.json(s);
});

app.post('/api/guilds/:guild_id/settings', (req, res) => {
  const id = String(req.params.guild_id);
  settingsStore[id] = req.body;
  saveSettings();
  res.json({ ok: true });
});

app.get('/api/guilds/:guild_id/channels', (req, res) => {
  const g = getGuildOr404(req.params.guild_id, res); if (!g) return;
  const out = g.channels.cache.filter(c => c.type === ChannelType.GuildText).map(c => ({ id: String(c.id), name: c.name }));
  res.json(out);
});

app.get('/api/guilds/:guild_id/roles', (req, res) => {
  const g = getGuildOr404(req.params.guild_id, res); if (!g) return;
  const everyoneId = g.roles.everyone.id;
  const out = g.roles.cache.filter(r => r.id !== everyoneId).map(r => ({ id: String(r.id), name: r.name }));
  res.json(out);
});

app.get('/', (req, res) => res.sendFile(path.join(__dirname, 'index.html')));

const token = process.env.DISCORD_TOKEN;
if (!token) { console.error('DISCORD_TOKEN not set'); process.exit(1); }

app.listen(8000, () => console.log('http://localhost:8000/'));
client.login(token).catch(err => { console.error(err); process.exit(1); });
async function ensureMuteRole(guild) {
  const cfg = settingsStore[String(guild.id)] || defaultSettings();
  let roleId = cfg.moderation && cfg.moderation.mute_role_id || null;
  let role = roleId ? guild.roles.cache.get(String(roleId)) : null;
  if (!role) {
    role = guild.roles.cache.find(r=>r.name.toLowerCase()==='muted');
  }
  if (!role) {
    try { role = await guild.roles.create({ name: 'Muted', color: 0x808080, reason: 'Mute role' }); } catch (_) { return null; }
  }
  cfg.moderation = cfg.moderation || {};
  cfg.moderation.mute_role_id = role.id;
  settingsStore[String(guild.id)] = cfg;
  saveSettings();
  return role.id;
}

File: discord_bot.py
import os
import asyncio
import json
from aiohttp import web
import discord
from discord.ext import commands
from discord import app_commands

PORT = int(os.environ.get("DASHBOARD_PORT", "8000"))
DISABLE_PRIV = os.environ.get("DISABLE_PRIVILEGED_INTENTS", "0") == "1"
BASE_URL = os.environ.get("DASHBOARD_BASE", f"http://localhost:{PORT}/")

intents = discord.Intents.none()
intents.guilds = True
intents.guild_messages = True
intents.message_content = not DISABLE_PRIV
intents.members = not DISABLE_PRIV

bot = commands.Bot(command_prefix="!", intents=intents)
LANG = "ko"
BOT_INSTANCES: dict[str, dict] = {}

async def api_stats(request):
    guild_count = len(bot.guilds)
    channel_count = sum(len(g.channels) for g in bot.guilds)
    member_count = 0
    online_count = 0
    for g in bot.guilds:
        try:
            member_count += g.member_count or 0
        except Exception:
            pass
        try:
            online_count += sum(1 for m in g.members if m.status and str(m.status) != "offline")
        except Exception:
            pass
    return web.json_response({
        "guilds": guild_count,
        "channels": channel_count,
        "members": member_count,
        "online": online_count
    })

async def api_guilds(request):
    data = [{"id": str(g.id), "name": g.name} for g in bot.guilds]
    return web.json_response({"guilds": data})

async def api_commands(request):
    cmds = [
        {"name": "핑", "type": "slash"},
        {"name": "대시보드", "type": "slash"},
        {"name": "공지", "type": "slash"},
        {"name": "서버정보", "type": "slash"},
        {"name": "유저정보", "type": "slash"}
    ]
    return web.json_response({"commands": cmds})

async def api_announce(request):
    try:
        payload = await request.json()
    except Exception:
        payload = {}
    channel_id = str(payload.get("channel_id", ""))
    message = str(payload.get("message", ""))
    if not channel_id or not message:
        return web.json_response({"ok": False, "error": "missing_params"}, status=400)
    channel = bot.get_channel(int(channel_id))
    if channel is None:
        return web.json_response({"ok": False, "error": "channel_not_found"}, status=404)
    try:
        await channel.send(message)
        return web.json_response({"ok": True})
    except Exception as e:
        return web.json_response({"ok": False, "error": str(e)}, status=500)

async def api_lang_get(request):
    return web.json_response({"lang": LANG})

async def api_lang_set(request):
    try:
        payload = await request.json()
    except Exception:
        payload = {}
    value = str(payload.get("lang", "")).lower()
    if value not in ("ko", "en"):
        return web.json_response({"ok": False}, status=400)
    global LANG
    LANG = value
    return web.json_response({"ok": True, "lang": LANG})

async def index_handler(request):
    return web.FileResponse(path=os.path.join(os.getcwd(), "index.html"))

def create_app():
    app = web.Application()
    app.router.add_get("/", index_handler)
    app.router.add_get("/api/stats", api_stats)
    app.router.add_get("/api/guilds", api_guilds)
    app.router.add_get("/api/commands", api_commands)
    app.router.add_post("/api/announce", api_announce)
    app.router.add_get("/api/lang", api_lang_get)
    app.router.add_post("/api/lang", api_lang_set)
    async def api_base(request):
        return web.json_response({"base": BASE_URL})
    app.router.add_get("/api/base", api_base)
    async def api_bots(request):
        items = []
        for bid, info in BOT_INSTANCES.items():
            items.append({"id": bid, "name": info.get("name", ""), "prefix": info.get("prefix", "!")})
        return web.json_response({"bots": items})
    async def api_bot_start(request):
        try:
            payload = await request.json()
        except Exception:
            payload = {}
        name = str(payload.get("name", "")).strip() or "bot"
        token = str(payload.get("token", "")).strip()
        prefix = str(payload.get("prefix", "!")).strip() or "!"
        if not token:
            return web.json_response({"ok": False}, status=400)
        bid = await start_additional_bot(name, token, prefix)
        return web.json_response({"ok": True, "id": bid})
    async def api_bot_stop(request):
        try:
            payload = await request.json()
        except Exception:
            payload = {}
        bid = str(payload.get("id", "")).strip()
        if not bid or bid not in BOT_INSTANCES:
            return web.json_response({"ok": False}, status=404)
        await stop_additional_bot(bid)
        return web.json_response({"ok": True})
    app.router.add_get("/api/bots", api_bots)
    app.router.add_post("/api/bot/start", api_bot_start)
    app.router.add_post("/api/bot/stop", api_bot_stop)
    static_dir = os.path.join(os.getcwd(), "static")
    if not os.path.isdir(static_dir):
        os.makedirs(static_dir, exist_ok=True)
    app.router.add_static("/static/", static_dir, show_index=False)
    return app

@bot.tree.command(name="핑")
async def cmd_ping(interaction: discord.Interaction):
    await interaction.response.send_message("퐁", ephemeral=True)

@bot.tree.command(name="대시보드")
async def cmd_dashboard(interaction: discord.Interaction):
    await interaction.response.send_message(BASE_URL, ephemeral=True)

@bot.tree.command(name="공지")
@app_commands.describe(채널="보낼 채널", 내용="메시지")
async def cmd_announce(interaction: discord.Interaction, 채널: discord.TextChannel, 내용: str):
    try:
        await 채널.send(내용)
        await interaction.response.send_message("전송 완료", ephemeral=True)
    except Exception as e:
        await interaction.response.send_message("오류", ephemeral=True)

@bot.tree.command(name="서버정보")
async def cmd_server(interaction: discord.Interaction):
    g = interaction.guild
    if not g:
        await interaction.response.send_message("서버에서만 사용", ephemeral=True)
        return
    await interaction.response.send_message(f"서버: {g.name} • 멤버: {g.member_count}", ephemeral=True)

@bot.tree.command(name="유저정보")
@app_commands.describe(사용자="대상 유저")
async def cmd_user(interaction: discord.Interaction, 사용자: discord.Member | None = None):
    m = 사용자 or interaction.user
    await interaction.response.send_message(f"유저: {m.display_name} • ID: {m.id}", ephemeral=True)

@bot.tree.command(name="경고")
@app_commands.describe(사용자="대상 유저", 사유="사유")
async def cmd_warn(interaction: discord.Interaction, 사용자: discord.Member, 사유: str):
    if not interaction.user.guild_permissions.manage_messages:
        await interaction.response.send_message("권한 없음", ephemeral=True)
        return
    try:
        await 사용자.send(f"경고: {사유}")
    except Exception:
        pass
    await interaction.response.send_message("경고 처리", ephemeral=True)

@bot.tree.command(name="뮤트")
@app_commands.describe(사용자="대상 유저", 분="분")
async def cmd_mute(interaction: discord.Interaction, 사용자: discord.Member, 분: int):
    if not interaction.user.guild_permissions.moderate_members:
        await interaction.response.send_message("권한 없음", ephemeral=True)
        return
    try:
        await 사용자.timeout(discord.utils.utcnow() + discord.timedelta(minutes=max(1, 분)))
        await interaction.response.send_message("뮤트 처리", ephemeral=True)
    except Exception:
        await interaction.response.send_message("오류", ephemeral=True)

@bot.tree.command(name="언뮤트")
@app_commands.describe(사용자="대상 유저")
async def cmd_unmute(interaction: discord.Interaction, 사용자: discord.Member):
    if not interaction.user.guild_permissions.moderate_members:
        await interaction.response.send_message("권한 없음", ephemeral=True)
        return
    try:
        await 사용자.timeout(None)
        await interaction.response.send_message("언뮤트 처리", ephemeral=True)
    except Exception:
        await interaction.response.send_message("오류", ephemeral=True)

@bot.event
async def on_ready():
    try:
        await bot.tree.sync()
    except Exception:
        pass

async def main():
    token = os.environ.get("DISCORD_TOKEN", "")
    if not token:
        raise RuntimeError("DISCORD_TOKEN not set")
    app = create_app()
    runner = web.AppRunner(app)
    await runner.setup()
    site = web.TCPSite(runner, "0.0.0.0", PORT)
    await site.start()
    await bot.start(token)

def build_bot(prefix: str) -> commands.Bot:
    intents2 = discord.Intents.none()
    intents2.guilds = True
    intents2.guild_messages = True
    intents2.message_content = not DISABLE_PRIV
    intents2.members = not DISABLE_PRIV
    b = commands.Bot(command_prefix=prefix, intents=intents2)
    @b.tree.command(name="핑")
    async def _ping(interaction: discord.Interaction):
        await interaction.response.send_message("퐁", ephemeral=True)
    @b.tree.command(name="대시보드")
    async def _dash(interaction: discord.Interaction):
        await interaction.response.send_message(BASE_URL, ephemeral=True)
    @b.event
    async def on_ready():
        try:
            await b.tree.sync()
        except Exception:
            pass
    return b

async def start_additional_bot(name: str, token: str, prefix: str) -> str:
    b = build_bot(prefix)
    bid = f"bot-{len(BOT_INSTANCES)+1}"
    BOT_INSTANCES[bid] = {"name": name, "prefix": prefix, "bot": b}
    asyncio.create_task(b.start(token))
    return bid

async def stop_additional_bot(bid: str):
    info = BOT_INSTANCES.get(bid)
    if not info:
        return
    b: commands.Bot = info.get("bot")
    try:
        await b.close()
    except Exception:
        pass
    BOT_INSTANCES.pop(bid, None)

if __name__ == "__main__":
    asyncio.run(main())